src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/StackExchangeRedis2Plus/SessionCache.cs

```diff
@@ -111,25 +137,28 @@ public Func<ProfilingSession> GetProfilingSession()
                 }
 
                 // Don't want to save data to a session to a NoOp - no way to clean it up easily or reliably.
+                // Don't want to save to a Datastore segment - could be another Redis segment or something else.
                 var segment = transaction.CurrentSegment;
-                if (!segment.IsValid)
+                ProfilingSession session = null;
+
+                lock (segment)
                 {
-                    return null;
+                    if (segment.IsValid && !segment.IsDone && segment.GetCategory() != "Datastore")
```

Comment:
segment.IsValid and segment.GetCategory does not change over time, we might want to pull these checks outside of the lock, and only lock around the conditional that checks that the segment is not done.